datalad/support/annexrepo.py

```diff
@@ -3590,9 +3590,25 @@ def _is_annex_work_tree_message(out):
             out,
             re.MULTILINE | re.DOTALL | re.IGNORECASE)
 
+
+class SafeDelCloseMixin(object):
+    """A helper class to use where __del__ would call .close() which might
+    fail if "too late in GC game"
+    """
+    def __del__(self):
+        try:
+            self.close()
+        except TypeError as exc:
+            if os.fdopen is None or lgr.debug is None:
+                # if we are late in the game and things already gc'ed in py3,
+                # it is Ok
+                return
+            raise
+
+
 # TODO: Why was this commented out?
 # @auto_repr
-class BatchedAnnexes(dict):
+class BatchedAnnexes(dict, SafeDelCloseMixin):
```

Comment:
It doesn't matter in this particular case, but the base class `dict` should be the right-most argument.

Reply:
thanks for the comments -- amended an force pushed!  Need to learn why `dict` should be the rightmost

Reply:
Oh no, hopefully on further reflection you don't disagree with my statement then :]

I'm making the claim that `SafeDelCloseMixin` should have higher precedence in the resolution chain, even if by design it is not intended define any methods that overlap with `dict`.

```py
class M():
    def blah(self):
        return "M"


class A():
    def blah(self):
        return "A"


class B0(A, M):
    pass


class B1(M, A):
    pass


print(B0.__mro__)  # (<class '__main__.B0'>, <class '__main__.A'>, <class '__main__.M'>, <class 'object'>)
print(B0().blah()) # A
print(B1.__mro__)  # (<class '__main__.B1'>, <class '__main__.M'>, <class '__main__.A'>, <class 'object'>)
print(B1().blah()) # M
```

Reply:
;) so let's keep as is now ;) but I guess that is why I typically saw mixins last, since they would then get lower precedence, BUT still probably would be invoked within MRO deeper (typically we call super first while inheriting) even when multiple subclasses (re)define the same method